File: src/gui/HUD.lua
--- src/gui/HUD.lua
-- @class HUD
-- Draws the heart counter and coin counter
local HUD = {}
local Player = require("src.entities.Player")

-- Love 2D load function 
function HUD:load()
    HUD.coinCounter = {} -- Coin counter table
    HUD.coinCounter.x = love.graphics.getWidth() - 200 -- Coin counter x position in HUD
    HUD.coinCounter.y = 50 -- Coin counter y position in HUD
    HUD.coinCounter.image = love.graphics.newImage("assets/world/coin.png") -- Load coin image
    HUD.coinCounter.width = HUD.coinCounter.image:getWidth() -- Get width of coin image
    HUD.coinCounter.height = HUD.coinCounter.image:getHeight() -- Get height of coin image
    HUD.coinCounter.scale = 3 -- Set scaling to 300%
    HUD.publicPixelFont = love.graphics.newFont("assets/fonts/public-pixel-font.ttf", 40) -- Coin counter font

    HUD.heartCounter = {} -- Heart counter table
    HUD.heartCounter.x = 0 -- Heart counter x position in HUD   
    HUD.heartCounter.y = 50 -- Heart counter y position in HUD  
    HUD.heartCounter.image = love.graphics.newImage("assets/world/heart.png") -- Load heart image
    HUD.heartCounter.width = HUD.heartCounter.image:getWidth() -- Get width of heart image
    HUD.heartCounter.height = HUD.heartCounter.image:getHeight() -- Get height of heart image
    HUD.heartCounter.scale = 3 -- Set scaling to 300%
    HUD.heartCounter.spaceBetween = HUD.heartCounter.width * HUD.heartCounter.scale + 30 -- Set the spacing between the hearts
end

-- Love 2D update function
function HUD:update(dt)

end

-- Love 2D draw function
function HUD:draw()
    HUD:drawCoinCounter() -- Draw coin counter to screen
    HUD:drawHeartCounter() -- Draw heart counter to screen
end

function HUD:drawHeartCounter()
    -- Iterate over the current player health table
    for i = 1, Player.health.current, 1 do
        local x = HUD.heartCounter.x + HUD.heartCounter.spaceBetween * i
        local shadowOffset = 2

        -- Draw shadow for heart counter with an offset of 3
        love.graphics.setColor(0, 0, 0, 0.5)
        love.graphics.draw(HUD.heartCounter.image, x + shadowOffset, HUD.heartCounter.y + shadowOffset, 0, HUD.heartCounter.scale,
            HUD.heartCounter.scale)
        love.graphics.setColor(1, 1, 1, 1)

        -- Draw heart image
        love.graphics.draw(HUD.heartCounter.image, x, HUD.heartCounter.y, 0, HUD.heartCounter.scale, HUD.heartCounter.scale)
    end
end

--[[
    This function is responsible for drawing the coin image and coin count to the screen.
]]
function HUD:drawCoinCounter()
    -- Set new font
    love.graphics.setFont(HUD.publicPixelFont)

    local shadowOffset = 3

    -- Draw shadow for coin counter with an offset of 3
    love.graphics.setColor(0, 0, 0, 0.5)
    love.graphics.draw(HUD.coinCounter.image, HUD.coinCounter.x + shadowOffset, HUD.coinCounter.y + 3, 0, HUD.coinCounter.scale,
        HUD.coinCounter.scale)
    love.graphics.setColor(1, 1, 1, 1)

    -- Draw coin image
    love.graphics.draw(HUD.coinCounter.image, HUD.coinCounter.x, HUD.coinCounter.y, 0, HUD.coinCounter.scale, HUD.coinCounter.scale)

    -- Coin counter position in the GUI
    local x = HUD.coinCounter.x + HUD.coinCounter.width * HUD.coinCounter.scale + 10
    local y = HUD.coinCounter.y + HUD.coinCounter.height / 2 * HUD.coinCounter.scale - HUD.publicPixelFont:getHeight() / 2

    -- Create a shadow around the coin
    love.graphics.setColor(0, 0, 0, 0.5)

    -- Offset shadow
    love.graphics.print(Player.coinCount, x + 3, y + 3)
    love.graphics.setColor(1, 1, 1, 1)
    love.graphics.print(Player.coinCount, x, y)
end

return HUD

File: internal/gui/hud.go
package gui

import (
	"fmt"
	"image"
	"image/color"
	_ "image/png"
	"os"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/text/v2"
)

// HUD constants — scaled for 640×360 internal resolution.
const (
	hudHeartX       = 10.0
	hudHeartY       = 8.0
	hudScale        = 1.5
	hudShadowOffset = 1.0
	hudCoinShadow   = 1.0
	hudCoinMarginX  = 10.0 // margin from right edge
	hudHeartSpacing = 4.0  // gap between hearts
)

// HUD draws the heart counter and coin counter.
type HUD struct {
	heartImg *ebiten.Image
	heartW   float64
	heartH   float64

	coinImg *ebiten.Image
	coinW   float64
	coinH   float64

	font *text.GoTextFace

	screenW float64
}

// NewHUD creates the HUD, loading heart/coin images and the pixel font.
func NewHUD(screenW, screenH float64) (*HUD, error) {
	heartImg, heartW, heartH, err := loadGUIImage("assets/world/heart.png")
	if err != nil {
		return nil, fmt.Errorf("loading heart: %w", err)
	}

	coinImg, coinW, coinH, err := loadGUIImage("assets/world/coin.png")
	if err != nil {
		return nil, fmt.Errorf("loading coin: %w", err)
	}

	fontSrc, err := LoadFont("assets/fonts/public-pixel-font.ttf")
	if err != nil {
		return nil, fmt.Errorf("loading HUD font: %w", err)
	}

	face := &text.GoTextFace{
		Source: fontSrc,
		Size:   FontMedium,
	}

	return &HUD{
		heartImg: heartImg,
		heartW:   heartW,
		heartH:   heartH,
		coinImg:  coinImg,
		coinW:    coinW,
		coinH:    coinH,
		font:     face,
		screenW:  screenW,
	}, nil
}

// Draw renders the HUD (hearts + coin counter).
func (h *HUD) Draw(screen *ebiten.Image, health, maxHealth, coinCount int) {
	h.drawHearts(screen, health)
	h.drawCoinCounter(screen, coinCount)
}

func (h *HUD) drawHearts(screen *ebiten.Image, health int) {
	scaledW := h.heartW * hudScale

	for i := 0; i < health; i++ {
		x := hudHeartX + float64(i)*(scaledW+hudHeartSpacing)

		// Shadow
		opts := &ebiten.DrawImageOptions{}
		opts.GeoM.Scale(hudScale, hudScale)
		opts.GeoM.Translate(x+hudShadowOffset, hudHeartY+hudShadowOffset)
		opts.ColorScale.Scale(0, 0, 0, 0.5)
		screen.DrawImage(h.heartImg, opts)

		// Heart
		opts2 := &ebiten.DrawImageOptions{}
		opts2.GeoM.Scale(hudScale, hudScale)
		opts2.GeoM.Translate(x, hudHeartY)
		screen.DrawImage(h.heartImg, opts2)
	}
}

func (h *HUD) drawCoinCounter(screen *ebiten.Image, coinCount int) {
	// Measure text to compute total width
	countStr := fmt.Sprintf("%d", coinCount)
	textW, textH := MeasureText(countStr, h.font)

	scaledCoinW := h.coinW * hudScale
	scaledCoinH := h.coinH * hudScale
	gap := 4.0 // gap between coin icon and text

	// Total width = coin + gap + text
	totalW := scaledCoinW + gap + textW

	// Anchor from the right edge
	coinX := h.screenW - hudCoinMarginX - totalW

	// Coin shadow
	opts := &ebiten.DrawImageOptions{}
	opts.GeoM.Scale(hudScale, hudScale)
	opts.GeoM.Translate(coinX+hudCoinShadow, hudHeartY+hudCoinShadow)
	opts.ColorScale.Scale(0, 0, 0, 0.5)
	screen.DrawImage(h.coinImg, opts)

	// Coin
	opts2 := &ebiten.DrawImageOptions{}
	opts2.GeoM.Scale(hudScale, hudScale)
	opts2.GeoM.Translate(coinX, hudHeartY)
	screen.DrawImage(h.coinImg, opts2)

	// Text — vertically centered with coin
	textX := coinX + scaledCoinW + gap
	textY := hudHeartY + scaledCoinH/2 - textH/2

	DrawTextShadow(screen, countStr, textX, textY, h.font, hudCoinShadow, color.White)
}

// loadGUIImage loads a PNG and returns the image + dimensions.
func loadGUIImage(path string) (*ebiten.Image, float64, float64, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, 0, 0, fmt.Errorf("opening %s: %w", path, err)
	}
	defer f.Close()

	img, _, err := image.Decode(f)
	if err != nil {
		return nil, 0, 0, fmt.Errorf("decoding %s: %w", path, err)
	}

	eimg := ebiten.NewImageFromImage(img)
	w := float64(eimg.Bounds().Dx())
	h := float64(eimg.Bounds().Dy())
	return eimg, w, h, nil
}
